File: backend/src/battle-rules/ability-engine.ts
import { emptyAbilityEffect } from "./ability-effect.js";
import { opponentOf, withStatusApplied } from "./battle-state.js";
import type { CombinationBook } from "./combination-book.js";
import type { CombinationModifier } from "./combination-modifiers.js";
import type { Mutation } from "./mutations.js";
import { playTurn } from "./turn-engine.js";
import type { BattleState, TurnAction, TurnResult } from "./types.js";

/**
 * Mirrors AbilityEngine.useAbility in battle_engine: plays `action`'s
 * elements as a normal turn (via playTurn — combinations, combo damage,
 * status ticks and the win condition already resolved there), then applies
 * whatever the actor's currently granted `mutations` add on top: a field
 * effect, and each status on whichever side it targets (most, like
 * Combustão, land on the opponent; a defensive one like Guarda targets the
 * actor instead — see StatusTarget in ability-effect.ts).
 * `combinationModifiers` passes straight through to playTurn.
 */
export function useAbility(
  state: BattleState,
  action: TurnAction,
  combinationBook: CombinationBook,
  mutations: readonly Mutation[],
  combinationModifiers: readonly CombinationModifier[] = [],
): TurnResult {
  const turnResult = playTurn(state, action, combinationBook, combinationModifiers);

  let effect = emptyAbilityEffect;
  for (const mutation of mutations) {
    effect = mutation.apply(effect);
  }

  let nextState = turnResult.state;
  if (effect.fieldEffect) {
    nextState = {
      ...nextState,
      activeFieldEffects: [...nextState.activeFieldEffects, effect.fieldEffect],
    };
  }
  if (effect.statusesToApply.length > 0) {
    const opponentId = opponentOf(state, action.actorId);
    for (const targeted of effect.statusesToApply) {
      const targetId = targeted.target === "actor" ? action.actorId : opponentId;
      nextState = withStatusApplied(nextState, targetId, targeted.status);
    }
  }

  return { state: nextState, triggeredCombinationId: turnResult.triggeredCombinationId };
}

File: packages/battle_engine/lib/src/ability_engine.dart
import 'ability.dart';
import 'ability_effect.dart';
import 'ability_result.dart';
import 'battle_state.dart';
import 'combatant.dart';
import 'combination_modifier.dart';
import 'targeted_status.dart';
import 'turn_action.dart';
import 'turn_engine.dart';

/// Resolves the use of an [Ability]: plays its base elements as a normal
/// turn (via [TurnEngine], so combinations still trigger, combo damage and
/// the win condition are already resolved, and the turn still passes),
/// then applies whatever its mutations add — a field effect, and any
/// [TargetedStatus]es on whichever side each one targets (most, like
/// Combustão, land on the opponent; a defensive one like Guarda targets
/// the actor instead — see [StatusTarget]). [AbilityEffect.hitCount] and
/// [AbilityEffect.critChanceBonus] are still not consumed by anything;
/// they're carried through for a future combat system to read.
class AbilityEngine {
  final TurnEngine turnEngine;

  const AbilityEngine(this.turnEngine);

  /// [combinationModifiers] is the actor's build-level extension point over
  /// a triggered combination's result (see [TurnEngine.playTurn]) —
  /// typically `build.combinationModifiers`.
  AbilityResult useAbility(
    BattleState state,
    Combatant actor,
    Ability ability, {
    List<CombinationModifier> combinationModifiers = const [],
  }) {
    final turnResult = turnEngine.playTurn(
      state,
      TurnAction(actor: actor, elements: ability.baseElements),
      combinationModifiers: combinationModifiers,
    );

    var effect = const AbilityEffect();
    for (final mutation in ability.mutations) {
      effect = mutation.apply(effect);
    }

    var nextState = turnResult.state;
    if (effect.fieldEffect != null) {
      nextState = nextState.withFieldEffect(effect.fieldEffect!);
    }
    if (effect.statusesToApply.isNotEmpty) {
      final opponent = state.opponentOf(actor);
      for (final targeted in effect.statusesToApply) {
        final target =
            targeted.target == StatusTarget.actor ? actor : opponent;
        nextState = nextState.withStatusApplied(target, targeted.status);
      }
    }

    return AbilityResult(
      state: nextState,
      effect: effect,
      triggeredCombination: turnResult.triggeredCombination,
    );
  }
}
